## Importamos las librerias necesarias

In [2]:
import pandas as pd

## Leemos el fichero csv de Git

In [3]:
url="https://raw.githubusercontent.com/Violainedlst/proyecto-da-promo-H-modulo-3-team-2-People-pulse/main/HR%20RAW%20DATA.csv"
data = pd.read_csv(url)

## Sacamos la información de csv para ver los tipos de datos de cada columna

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1614 entries, 0 to 1613
Data columns (total 42 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Unnamed: 0                1614 non-null   int64  
 1   Age                       1614 non-null   object 
 2   Attrition                 1614 non-null   object 
 3   BusinessTravel            842 non-null    object 
 4   DailyRate                 1614 non-null   object 
 5   Department                302 non-null    object 
 6   DistanceFromHome          1614 non-null   int64  
 7   Education                 1614 non-null   int64  
 8   EducationField            869 non-null    object 
 9   employeecount             1614 non-null   int64  
 10  employeenumber            1183 non-null   object 
 11  EnvironmentSatisfaction   1614 non-null   int64  
 12  Gender                    1614 non-null   int64  
 13  HourlyRate                1614 non-null   object 
 14  JobInvol

## Recorremos las columnas para ver la información que contiene cada una

In [5]:
for column in data.columns:
  
    print(f"Columna: {column}")
    print(data[column].unique())

    print("\n--------------------\n")

Columna: Unnamed: 0
[   0    1    2 ... 1611 1612 1613]

--------------------

Columna: Age
['51' '52' '42' '47' '46' '48' '59' '41' '56' '38' '55' '40' '58' '35'
 '45' '33' '36' '34' 'forty-seven' '53' '43' '60' '32' '37' '49' '39' '50'
 '44' '30' 'fifty-eight' '29' '31' '54' '57' '27' 'thirty-six' '28' '26'
 'fifty-five' '25' 'fifty-two' 'thirty-one' '24' 'thirty' '23' '22' '21'
 '20' 'twenty-six' '19' 'thirty-seven' '18' 'thirty-two' 'twenty-four']

--------------------

Columna: Attrition
['No' 'Yes']

--------------------

Columna: BusinessTravel
[nan 'travel_rarely' 'travel_frequently' 'non-travel']

--------------------

Columna: DailyRate
['684,0$' '699,0$' '532,0$' '359,0$' '1319,0$' '117,0$' '1435,0$' '635,0$'
 '1276,0$' '840,0$' '247,0$' '1369,0$' '201,0$' '1360,0$' '692,0$'
 '1398,0$' '286,0$' '1402,0$' '819,0$' '884,0$' '1238,0$' '515,0$'
 '1223,0$' '202,0$' '928,0$' '607,0$' '266,0$' '429,0$' '589,0$' 'nan$'
 '1180,0$' '1282,0$' '776,0$' '665,0$' '526,0$' '1034,0$' '1403,

## Empezamos a limpiar los datos por columnas


In [6]:
# Convertimos todos los nombres de las columnas a minúsculas
data.columns = [column.lower() for column in data.columns]

# Modificamos los nombres de las columnas para mejor comprensión lectora
data.rename(columns={
    'unnamed: 0': 'id',
    'age': 'age',
    'attrition': 'attrition',
    'businesstravel': 'business_travel',
    'dailyrate': 'daily_rate',
    'department': 'department',
    'distancefromhome': 'distance_from_home',
    'education': 'education',
    'educationfield': 'education_field',
    'employeecount': 'employee_count',
    'employeenumber': 'employee_number',
    'environmentsatisfaction': 'environment_satisfaction',
    'gender': 'gender',
    'hourlyrate': 'hourly_rate',
    'jobinvolvement': 'job_involvement',
    'joblevel': 'job_level',
    'jobrole': 'job_role',
    'jobsatisfaction': 'job_satisfaction',
    'maritalstatus': 'marital_status',
    'monthlyincome': 'monthly_income',
    'monthlyrate': 'monthly_rate',
    'numcompaniesworked': 'num_companies_worked',
    'over18': 'over_18',
    'overtime': 'overtime',
    'percentsalaryhike': 'percent_salary_hike',
    'performancerating': 'performance_rating',
    'relationshipsatisfaction': 'relationship_satisfaction',
    'standardhours': 'standard_hours',
    'stockoptionlevel': 'stock_option_level',
    'totalworkingyears': 'total_working_years',
    'trainingtimeslastyear': 'training_times_last_year',
    'worklifebalance': 'work_life_balance',
    'yearsatcompany': 'years_at_company',
    'yearsincurrentrole': 'years_in_current_role',
    'yearssincelastpromotion': 'years_since_last_promotion',
    'yearswithcurrmanager': 'years_with_curr_manager',
    'sameasmonthlyincome': 'same_as_monthly_income',
    'datebirth': 'date_birth',
    'salary': 'salary',
    'roledepartament': 'role_department',
    'numberchildren': 'number_children',
    'remotework': 'remote_work'
}, inplace=True)

In [7]:
data

id age attrition business_travel daily_rate                department  \
0        0  51        No             NaN     684,0$                       NaN   
1        1  52        No             NaN     699,0$                       NaN   
2        2  42        No   travel_rarely     532,0$   Research & Development    
3        3  47        No   travel_rarely     359,0$                       NaN   
4        4  46        No             NaN    1319,0$                       NaN   
...    ...  ..       ...             ...        ...                       ...   
1609  1609  36       Yes   travel_rarely     530,0$                       NaN   
1610  1610  45        No      non-travel     805,0$                       NaN   
1611  1611  39        No   travel_rarely     903,0$                       NaN   
1612  1612  36        No      non-travel    1229,0$                       NaN   
1613  1613  46        No             NaN     566,0$                       NaN   

      distance_from_home  education   education_field  employee_count  ...  \
0                      6          3               NaN               1  ...   
1                      1          4     Life Sciences               1  ...   
2                      4          2  Technical Degree               1  ...   
3                      2          4           Medical               1  ...   
4                      3          3  Technical Degree               1  ...   
...                  ...        ...               ...             ...  ...   
1609                   3          1     Life Sciences               1  ...   
1610                   4          2               NaN               1  ...   
1611                 -13          5               NaN               1  ...   
1612                   8          4  Technical Degree               1  ...   
1613                   7          2           Medical               1  ...   

     years_at_company  years_in_current_role  years_since_last_promotion  \
0                  20                    NaN                          15   
1                  33                    NaN                          11   
2                  22                    NaN                          11   
3                  20                    NaN                           5   
4                  19                    NaN                           2   
...               ...                    ...                         ...   
1609               16                    NaN                           3   
1610                9                    NaN                           0   
1611                8                    NaN                           0   
1612                7                    NaN                           0   
1613                8                    NaN                           0   

     years_with_curr_manager  same_as_monthly_income  date_birth       salary  \
0                         15                 19537,0        1972  1000000000$   
1                          9                 19999,0        1971  1000000000$   
2                         15                 19232,0        1981  1000000000$   
3                          6                 17169,0        1976  1000000000$   
4                          8                     NaN        1977  1000000000$   
...                      ...                     ...         ...          ...   
1609                       7                 10325,0        1987  1000000000$   
1610                       8                  4447,0        1978  1000000000$   
1611                       7                     NaN        1984  1000000000$   
1612                       7                     NaN        1987  1000000000$   
1613                       7                 10845,0        1977  1000000000$   

                           role_department number_children remote_work  
0                                      NaN             NaN         Yes  
1                                      NaN             NaN        

In [8]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1614 entries, 0 to 1613
Data columns (total 42 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   id                          1614 non-null   int64  
 1   age                         1614 non-null   object 
 2   attrition                   1614 non-null   object 
 3   business_travel             842 non-null    object 
 4   daily_rate                  1614 non-null   object 
 5   department                  302 non-null    object 
 6   distance_from_home          1614 non-null   int64  
 7   education                   1614 non-null   int64  
 8   education_field             869 non-null    object 
 9   employee_count              1614 non-null   int64  
 10  employee_number             1183 non-null   object 
 11  environment_satisfaction    1614 non-null   int64  
 12  gender                      1614 non-null   int64  
 13  hourly_rate                 1614 

## Decidimos entre Silvia P, Diana y Mábel, quitar columnas id_employee por no encontrarle utilidad

In [9]:
# Eliminar las columnas 'id' y 'same_as_monthly_income'
data_noid = data.drop(columns=["id", "same_as_monthly_income"])

# Mostrar el DataFrame resultante
data_noid

age attrition business_travel daily_rate                department  \
0     51        No             NaN     684,0$                       NaN   
1     52        No             NaN     699,0$                       NaN   
2     42        No   travel_rarely     532,0$   Research & Development    
3     47        No   travel_rarely     359,0$                       NaN   
4     46        No             NaN    1319,0$                       NaN   
...   ..       ...             ...        ...                       ...   
1609  36       Yes   travel_rarely     530,0$                       NaN   
1610  45        No      non-travel     805,0$                       NaN   
1611  39        No   travel_rarely     903,0$                       NaN   
1612  36        No      non-travel    1229,0$                       NaN   
1613  46        No             NaN     566,0$                       NaN   

      distance_from_home  education   education_field  employee_count  \
0                      6          3               NaN               1   
1                      1          4     Life Sciences               1   
2                      4          2  Technical Degree               1   
3                      2          4           Medical               1   
4                      3          3  Technical Degree               1   
...                  ...        ...               ...             ...   
1609                   3          1     Life Sciences               1   
1610                   4          2               NaN               1   
1611                 -13          5               NaN               1   
1612                   8          4  Technical Degree               1   
1613                   7          2           Medical               1   

     employee_number  ...  work_life_balance  years_at_company  \
0              162,0  ...                3,0                20   
1              259,0  ...                3,0                33   
2              319,0  ...                NaN                22   
3                NaN  ...                NaN                20   
4                NaN  ...                3,0                19   
...              ...  ...                ...               ...   
1609           967,0  ...                3,0                16   
1610           972,0  ...                2,0                 9   
1611             NaN  ...                3,0                 8   
1612           990,0  ...                3,0                 7   
1613             NaN  ...                3,0                 8   

     years_in_current_role  years_since_last_promotion  \
0                      NaN                          15   
1                      NaN                          11   
2                      NaN                          11   
3                      NaN                           5   
4                      NaN                           2   
...                    ...                         ...   
1609                   NaN                           3   
1610                   NaN                           0   
1611                   NaN                           0   
1612                   NaN                           0   
1613                   NaN                           0   

      years_with_curr_manager date_birth       salary  \
0                          15       1972  1000000000$   
1                           9       1971  1000000000$   
2                          15       1981  1000000000$   
3                           6       1976  1000000000$   
4                           8       1977  1000000000$   
...                       ...        ...          ...   
1609                        7       1987  1000000000$   
1610                        8       1978  1000000000$   
1611                        7       1984  1000000000$   
1612                        7       1987  1000000000$   
1613                        7       1977  1000000000$   

                           role_department number_ch

### Lo que hay que cambiar de la 0 a la 21

- `daily_rate`: quitar símbolo `$` y convertir a `int`

In [10]:
# Aseguramos que los valores de 'daily_rate' son de tipo string
data_noid['daily_rate'] = data_noid['daily_rate'].astype(str)

# Lista de símbolos a eliminar en los valores de la columna
quitar_simbolos = [' ', '$']

# Itera sobre cada símbolo en la lista de símbolos a eliminar
for simbolo in quitar_simbolos:
    # Elimina el símbolo actual de todos los valores de la columna 'daily_rate'
    data_noid['daily_rate'] = data_noid['daily_rate'].str.replace(simbolo, '', regex=False)

# Rellena los valores NaN y los valores vacíos con '0'
data_noid['daily_rate'] = data_noid['daily_rate'].replace(['nan', ''], '0')

# Convierte la columna 'daily_rate' a tipo int, manejando errores
data_noid['daily_rate'] = pd.to_numeric(data_noid['daily_rate'], errors='coerce').fillna(0).astype(int)

# Muestra las primeras filas del DataFrame para verificar que los valores de la columna se han modificado correctamente
data_noid.head(2)



age attrition business_travel  daily_rate department  distance_from_home  \
0  51        No             NaN           0        NaN                   6   
1  52        No             NaN           0        NaN                   1   

   education education_field  employee_count employee_number  ...  \
0          3             NaN               1           162,0  ...   
1          4   Life Sciences               1           259,0  ...   

   work_life_balance  years_at_company years_in_current_role  \
0                3,0                20                   NaN   
1                3,0                33                   NaN   

   years_since_last_promotion  years_with_curr_manager date_birth  \
0                          15                       15       1972   
1                          11                        9       1971   

        salary role_department number_children  remote_work  
0  1000000000$             NaN             NaN          Yes  
1  1000000000$             NaN             NaN            1  

[2 rows x 40 columns]

- `age` es `object` y debería ser `int`
- `hourly_rate` aparece como `object` y debería ser `int`
- `monthly_income`: convertir a `int`

In [11]:
columnas3 = ['age', 'hourly_rate', 'monthly_income']

# Itera sobre cada columna en la lista de columnas a convertir
for columna in columnas3:
    # Asegúrate de que los valores de la columna son de tipo string
    data_noid[columna] = data_noid[columna].astype(str)
    # Elimina cualquier valor no numérico o vacío y reemplaza por 0
    data_noid[columna] = data_noid[columna].replace(['nan', '', ' '], '0')
    # Convierte la columna a tipo numérico (int)
    data_noid[columna] = pd.to_numeric(data_noid[columna], errors='coerce').fillna(0).astype(int)

# Muestra las primeras filas del DataFrame para verificar que los valores de las columnas se han modificado correctamente
data_noid.head(2)

age attrition business_travel  daily_rate department  distance_from_home  \
0   51        No             NaN           0        NaN                   6   
1   52        No             NaN           0        NaN                   1   

   education education_field  employee_count employee_number  ...  \
0          3             NaN               1           162,0  ...   
1          4   Life Sciences               1           259,0  ...   

   work_life_balance  years_at_company  years_in_current_role  \
0                3,0                20                    NaN   
1                3,0                33                    NaN   

   years_since_last_promotion  years_with_curr_manager date_birth  \
0                          15                       15       1972   
1                          11                        9       1971   

        salary role_department  number_children  remote_work  
0  1000000000$             NaN              NaN          Yes  
1  1000000000$             NaN              NaN            1  

[2 rows x 40 columns]

- `employee_number` sale como `float` y debería ser `int`

In [12]:
# Eliminar caracteres no numéricos de 'employee_number'
data_noid['employee_number'] = data_noid['employee_number'].astype(str).str.replace('[^0-9.]', '', regex=True)

# Rellenar los valores NaN con 0
data_noid['employee_number'] = pd.to_numeric(data_noid['employee_number'], errors='coerce').fillna(0)

# Convertir la columna 'employee_number' de float a int
data_noid['employee_number'] = data_noid['employee_number'].astype(int)

# Mostrar los primeros registros para verificar
data_noid.head(2)

age attrition business_travel  daily_rate department  distance_from_home  \
0   51        No             NaN           0        NaN                   6   
1   52        No             NaN           0        NaN                   1   

   education education_field  employee_count  employee_number  ...  \
0          3             NaN               1             1620  ...   
1          4   Life Sciences               1             2590  ...   

   work_life_balance  years_at_company  years_in_current_role  \
0                3,0                20                    NaN   
1                3,0                33                    NaN   

   years_since_last_promotion  years_with_curr_manager date_birth  \
0                          15                       15       1972   
1                          11                        9       1971   

        salary role_department  number_children  remote_work  
0  1000000000$             NaN              NaN          Yes  
1  1000000000$             NaN              NaN            1  

[2 rows x 40 columns]

- cambiar todos los `valores` a `minúsculas`

In [13]:
# Lista de columnas a convertir a minúsculas de al 0 a la 21
columns_to_lower = [
    'age', 'attrition', 'business_travel', 'daily_rate', 'department', 
    'distance_from_home', 'education', 'education_field', 'employee_count', 
    'employee_number', 'environment_satisfaction', 'gender', 'hourly_rate', 
    'job_involvement', 'job_level', 'job_role', 'job_satisfaction', 
    'marital_status', 'monthly_income', 'monthly_rate'
]

# Convertir a minúsculas las columnas especificadas
for column in columns_to_lower:
    if data_noid[column].dtype == 'object':  # Verifica si la columna es de tipo str
        data_noid[column] = data_noid[column].str.lower()
        
data_noid.head(5)


age attrition business_travel  daily_rate                department  \
0   51        no             NaN           0                       NaN   
1   52        no             NaN           0                       NaN   
2   42        no   travel_rarely           0   research & development    
3   47        no   travel_rarely           0                       NaN   
4   46        no             NaN           0                       NaN   

   distance_from_home  education   education_field  employee_count  \
0                   6          3               NaN               1   
1                   1          4     life sciences               1   
2                   4          2  technical degree               1   
3                   2          4           medical               1   
4                   3          3  technical degree               1   

   employee_number  ...  work_life_balance  years_at_company  \
0             1620  ...                3,0                20   
1             2590  ...                3,0                33   
2             3190  ...                NaN                22   
3                0  ...                NaN                20   
4                0  ...                3,0                19   

   years_in_current_role  years_since_last_promotion  years_with_curr_manager  \
0                    NaN                          15                       15   
1                    NaN                          11                        9   
2                    NaN                          11                       15   
3                    NaN                           5                        6   
4                    NaN                           2                        8   

  date_birth       salary                       role_department  \
0       1972  1000000000$                                   NaN   
1       1971  1000000000$                                   NaN   
2       1981  1000000000$   ManaGER  -  Research & Development    
3       1976  1000000000$                                   NaN   
4       1977  1000000000$                                   NaN   

   number_children  remote_work  
0              NaN          Yes  
1              NaN            1  
2              NaN            1  
3              NaN        False  
4              NaN            0  

[5 rows x 40 columns]

- quitar los `NaN` de las columnas que lo tengan

In [14]:
# Reemplazar NaN en las columnas especificadas por 'No aplica'
columns_to_replace = ['business_travel', 'department', 'education_field', 'marital_status']
replacement_value = 'no aplica'

data_noid[columns_to_replace] = data_noid[columns_to_replace].fillna(replacement_value)

# Mostrar los primeros registros para verificar
data_noid.head(2)


age attrition business_travel  daily_rate department  distance_from_home  \
0   51        no       no aplica           0  no aplica                   6   
1   52        no       no aplica           0  no aplica                   1   

   education education_field  employee_count  employee_number  ...  \
0          3       no aplica               1             1620  ...   
1          4   life sciences               1             2590  ...   

   work_life_balance  years_at_company  years_in_current_role  \
0                3,0                20                    NaN   
1                3,0                33                    NaN   

   years_since_last_promotion  years_with_curr_manager date_birth  \
0                          15                       15       1972   
1                          11                        9       1971   

        salary role_department  number_children  remote_work  
0  1000000000$             NaN              NaN          Yes  
1  1000000000$             NaN              NaN            1  

[2 rows x 40 columns]

## Limpiamos las columnas de la 21 a la 31 (ambas incluidas)

Columnas verificadas donde no había nada que modificar:

- COLUMNA 21: "num_companies_worked" tiene 1614 entradas no nulas, como int64, hemos hecho un sample para doble verificación.

- COLUMNA 24: "percent_salary_hike" tiene 1614 entradas no nulas, como int64, hemos hecho un sample para doble verificación.

- COLUMNA 26: "relationship_satisfaction" tiene 1614 entradas no nulas, como int64, hemos hecho un sample para doble verificación.

- COLUMNA 28: "stock_option_level" tiene 1614 entradas no nulas, como int64, hemos hecho un sample para doble verificación.

- COLUMNA 30: "training_times_last_year" tiene 1614 entradas no nulas, como int64, hemos hecho un sample para doble verificación. 

- COLUMNA 32: "years_at_company" tiene 1614 entradas no nulas, como int64, hemos hecho un sample para doble verificación. 

- COLUMNA 34: "years_since_last_promotion" tiene 1614 entradas no nulas, como int64, hemos hecho un sample para doble verificación. 

- COLUMNA35: "years_with_curr_manager" tiene 1614 entradas no nulas, como int64, hemos hecho un sample para doble verificación. 

- COLUMNA36: "date_birth" tiene 1614 entradas no nulas, como int64, hemos hecho un sample para doble verificación. 

In [15]:
# Comprobación de la columna 21
data_noid["num_companies_worked"].sample(5)

1198    1
919     1
1005    1
443     4
234     1
Name: num_companies_worked, dtype: int64

In [16]:
# Comprobación de la columna 22
data_noid["over_18"].sample(5)

745     NaN
1010    NaN
827       Y
159       Y
810     NaN
Name: over_18, dtype: object

In [17]:
# LIMPIEZA DE LA COLUMNA 22 "over_18"
nuevos_valores = {"Y": "Yes", "N": "No"}
data_noid["over_18"] = data_noid["over_18"].map(nuevos_valores)
data_noid['over_18'] = data_noid['over_18'].fillna('N/A') #Usamos el "fillna" para cambiar el Nan (formato numpy) a N/A

In [18]:
# Comprobación de la columna 23
data_noid["overtime"].sample(5)

826      No
557      No
1091    NaN
709     NaN
1232     No
Name: overtime, dtype: object

In [19]:
# LIMPIEZA DE LA COLUMNA 23 "overtime"
data_noid['overtime'] = data_noid['overtime'].fillna('N/A') #Usamos el "fillna" para cambiar el Nan (formato numpy) a N/A

In [20]:
# Comprobación de la columna 24
data_noid["percent_salary_hike"].sample(5)

621    18
686    16
496    19
13     12
582    13
Name: percent_salary_hike, dtype: int64

In [21]:
# Comprobación de la columna 25
data_noid["performance_rating"].sample(5)

948    3,0
483    3,0
199    3,0
177    4,0
684    NaN
Name: performance_rating, dtype: object

In [22]:
# LIMPIEZA DE LA COLUMNA 25 "performance_rating"

# Nos aseguramos de que todos los valores son cadenas usando .astype(str)
data_noid["performance_rating"] = data_noid["performance_rating"].astype(str)

# Reemplazamos las comas por puntos
data_noid["performance_rating"] = data_noid["performance_rating"].str.replace(',', '.')

# Convertimos la columna a tipo numérico (dtype: "object" pasa a ser dtype: "float64")
data_noid["performance_rating"] = pd.to_numeric(data_noid["performance_rating"], errors='coerce')

# Cambiamos los valores que aparecen como NaN para darles valor 0
data_noid["performance_rating"] = data_noid["performance_rating"].fillna(0) 

# Cambiamos los valores de tipo string a int usando ".astype(int)"
data_noid["performance_rating"] = data_noid["performance_rating"].astype(int)

In [23]:
# Comprobación de la columna 26
data_noid["relationship_satisfaction"].sample(5)

771     3
1328    2
352     4
698     3
939     4
Name: relationship_satisfaction, dtype: int64

In [24]:
# Comprobación de la columna 27
data_noid["standard_hours"].sample(5)

131    80,0
455     NaN
448     NaN
23     80,0
389    80,0
Name: standard_hours, dtype: object

In [25]:
# LIMPIEZA DE LA COLUMNA 27 "standard_hours"

# Nos aseguramos de que todos los valores son cadenas usando .astype(str)
data_noid["standard_hours"] = data_noid["standard_hours"].astype(str)

# Reemplazamos las comas por puntos
data_noid["standard_hours"] = data_noid["standard_hours"].str.replace(',', '.')

# Convertimos la columna a tipo numérico (dtype: "object" pasa a ser dtype: "float64")
data_noid["standard_hours"] = pd.to_numeric(data_noid["standard_hours"], errors='coerce')

# Cambiamos los valores que aparecen como NaN para darles valor 0
data_noid["standard_hours"] = data_noid["standard_hours"].fillna(0) 

# Cambiamos los valores de tipo string a int usando ".astype(int)"
data_noid["standard_hours"] = data_noid["standard_hours"].astype(int)

In [26]:
# Comprobación de la columna 28
data_noid["stock_option_level"].sample(5)

1074    1
690     1
1350    0
958     1
920     0
Name: stock_option_level, dtype: int64

In [27]:
# Comprobación de la columna 29
data_noid["total_working_years"].sample(5)

1243     NaN
1221    10,0
1269    17,0
704      6,0
798      5,0
Name: total_working_years, dtype: object

In [28]:
# LIMPIEZA DE LA COLUMNA 29 "total_working_years"

# Nos aseguramos de que todos los valores son cadenas usando .astype(str)
data_noid["total_working_years"] = data_noid["total_working_years"].astype(str)

# Reemplazamos las comas por puntos
data_noid["total_working_years"] = data_noid["total_working_years"].str.replace(',', '.')

# Convertimos la columna a tipo numérico (dtype: "object" pasa a ser dtype: "float64")
data_noid["total_working_years"] = pd.to_numeric(data_noid["total_working_years"], errors='coerce')

# Cambiamos los valores que aparecen como NaN para dales valor 0
data_noid["total_working_years"] = data_noid["total_working_years"].fillna(0) 

# Cambiamos los valores de tipo string a int usando ".astype(int)"
data_noid["total_working_years"] = data_noid["total_working_years"].astype(int)

In [29]:
# Comprobación de la columna 30
data_noid["training_times_last_year"].sample(5)

1111    2
458     2
1481    5
1341    2
1410    4
Name: training_times_last_year, dtype: int64

In [30]:
# Comprobación de la columna 31
data_noid["work_life_balance"].sample(5)

583     3,0
30      4,0
474     3,0
1280    2,0
1491    3,0
Name: work_life_balance, dtype: object

In [31]:
# LIMPIEZA DE LA COLUMNA 31 "work_life_balance"

# Nos aseguramos de que todos los valores son cadenas usando .astype(str)
data_noid["work_life_balance"] = data_noid["work_life_balance"].astype(str)

# Reemplazamos las comas por puntos
data_noid["work_life_balance"] = data_noid["work_life_balance"].str.replace(',', '.')

# Convertimos la columna a tipo numérico (dtype: "object" pasa a ser dtype: "float64")
data_noid["work_life_balance"] = pd.to_numeric(data_noid["work_life_balance"], errors='coerce')

# Cambiamos los valores que aparecen como NaN para darles valor 0
data_noid["work_life_balance"] = data_noid["work_life_balance"].fillna(0) 

# Cambiamos los valores de tipo string a int usando ".astype(int)"
data_noid["work_life_balance"] = data_noid["work_life_balance"].astype(int)

In [45]:
data_noid.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1614 entries, 0 to 1613
Data columns (total 40 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   age                         1614 non-null   int64  
 1   attrition                   1614 non-null   object 
 2   business_travel             1614 non-null   object 
 3   daily_rate                  1614 non-null   int64  
 4   department                  1614 non-null   object 
 5   distance_from_home          1614 non-null   int64  
 6   education                   1614 non-null   int64  
 7   education_field             1614 non-null   object 
 8   employee_count              1614 non-null   int64  
 9   employee_number             1614 non-null   int64  
 10  environment_satisfaction    1614 non-null   int64  
 11  gender                      1614 non-null   int64  
 12  hourly_rate                 1614 non-null   int64  
 13  job_involvement             1614 

In [47]:
# Comprobación de la columna 32
data_noid["years_at_company"].sample(5)

561    10
380    10
228     9
236    10
818     3
Name: years_at_company, dtype: int64

In [51]:
# Comprobación de la columna 33
data_noid['years_in_current_role'] = data_noid['years_in_current_role'].fillna('N/A')

In [52]:
# Comprobación de la columna 34
data_noid["years_since_last_promotion"].sample(5)

537     1
1477    0
100     5
616     0
348     1
Name: years_since_last_promotion, dtype: int64

In [53]:
# Comprobación de la columna 35
data_noid["years_with_curr_manager"].sample(5)

1223    2
1462    0
936     3
1529    0
1183    2
Name: years_with_curr_manager, dtype: int64

In [54]:
# Comprobación de la columna 36
data_noid["date_birth"].sample(5)

1145    1978
1286    1981
1247    2000
629     1971
1483    1990
Name: date_birth, dtype: int64

In [66]:
# Comprobación de la columna 37 #Ver con equipo que quieren hacer
data_noid["salary"].sample(25)
data_noid["salary"] = data_noid["salary"].str.replace('1000000000$', 'Unknown')

In [70]:
# Comprobación de la columna 38

data_noid["role_department"] = data_noid["role_department"].fillna('N/A')

In [73]:
# Comprobación de la columna 39

data_noid["number_children"] = data_noid["number_children"].fillna('N/A')

In [80]:
# Comprobación de la columna 40
data_noid["remote_work"].sample(25)

#primero paso todo a lower para usar el lambda
data_noid['remote_work'] = data_noid['remote_work'].str.lower()

# Reemplazar valores que no sean "yes" o "no" con "no"
data_noid['remote_work'] = data_noid['remote_work'].apply(lambda x: 'yes' if x == 'yes' else 'no')


In [81]:
data_noid["remote_work"].sample(25)

726     yes
972      no
935      no
381     yes
1470     no
361      no
1379     no
116      no
598      no
1571     no
701      no
829      no
245      no
516      no
506      no
379      no
1090     no
210      no
475     yes
905      no
418      no
1355     no
1509     no
236      no
913      no
Name: remote_work, dtype: object